tpm2/src/util/nsl_util.c

```diff
@@ -0,0 +1,983 @@
+    return 1;
+  }
+  unsigned char *index = message;
+
+  // Parse out the nonce.
+  memcpy(nonce_len, index, sizeof(size_t));
+  index += sizeof(size_t);
+  *nonce = calloc(*nonce_len, sizeof(unsigned char));
+  if (*nonce == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the nonce buffer.");
+
+    *nonce_len = 0;
+
+    memset(message, 0, message_len);
+    free(message);
+
+    return 1;
+  }
+  memcpy(*nonce, index, *nonce_len);
+
+  memset(message, 0, message_len);
+  free(message);
+
+  return 0;
+}
+
+//
+// setup_public_evp_context
+//
+EVP_PKEY_CTX *setup_public_evp_context(const char *filepath)
+{
+  FILE *f = fopen(filepath, "r");
+
+  if (NULL == f)
+  {
+    kmyth_log(LOG_ERR, "Failed to open public key file.");
+    return NULL;
+  }
+
+  EVP_PKEY *pkey = PEM_read_PUBKEY(f, NULL, NULL, NULL);
+
+  if (NULL == pkey)
+  {
+    kmyth_log(LOG_ERR, "Failed to load public key file.");
+    if (fclose(f) != 0)
+    {
+      kmyth_log(LOG_ERR, "Failed to close public key file.");
+    }
+    return NULL;
+  }
+
+  if (fclose(f) != 0)
+  {
+    kmyth_log(LOG_ERR, "Failed to close public key file.");
+    EVP_PKEY_free(pkey);
+    return NULL;
+  }
+
+  EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(pkey, NULL);
+
+  if (NULL == ctx)
+  {
+    kmyth_log(LOG_ERR, "Failed to create public key EVP context.");
+    EVP_PKEY_free(pkey);
+    return NULL;
+  }
+
+  EVP_PKEY_free(pkey);
+
+  return ctx;
+}
+
+//
+// setup_private_evp_context()
+//
+EVP_PKEY_CTX *setup_private_evp_context(const char *filepath)
+{
+  FILE *f = fopen(filepath, "r");
+
+  if (NULL == f)
+  {
+    kmyth_log(LOG_ERR, "Failed to open private key file.");
+    return NULL;
+  }
+
+  EVP_PKEY *pkey = PEM_read_PrivateKey(f, NULL, NULL, NULL);
+
+  if (NULL == pkey)
+  {
+    kmyth_log(LOG_ERR, "Failed to load private key file.");
+    if (fclose(f) != 0)
+    {
+      kmyth_log(LOG_ERR, "Failed to close private key file.");
+    }
+    return NULL;
+  }
+
+  if (fclose(f) != 0)
+  {
+    kmyth_log(LOG_ERR, "Failed to close private key file.");
+    EVP_PKEY_free(pkey);
+    return NULL;
+  }
+
+  EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(pkey, NULL);
+
+  if (NULL == ctx)
+  {
+    kmyth_log(LOG_ERR, "Failed to create private key EVP context.");
+    EVP_PKEY_free(pkey);
+
+    return NULL;
+  }
+
+  EVP_PKEY_free(pkey);
+
+  return ctx;
+}
+
+//
+// generate_session_key()
+//
+int generate_session_key(unsigned char *nonce_a, size_t nonce_a_len,
+                         unsigned char *nonce_b, size_t nonce_b_len,
+                         unsigned char **key, size_t *key_len)
+{
+  // Build the combined nonce as the base for key generation.
+  size_t len = nonce_a_len + nonce_b_len;
+  unsigned char *nonces = calloc(len, sizeof(unsigned char));
+  unsigned char *index = nonces;
+
+  memcpy(index, nonce_a, nonce_a_len);
+  index += nonce_a_len;
+  memcpy(index, nonce_b, nonce_b_len);
+
+  // Setup the message digest context.
+  //const EVP_MD *type = EVP_sha3_512();
```

Comment:
At this point, I don't see any reason to leave the commented out code related to the SHA3-512 based approach, but if you think they might still be helpful, leaving them is OK.

Reply:
Yeah, we can remove this.